gnoi/gnoi.go

```diff
@@ -745,24 +752,642 @@ func (s *system) getSupervisorRole(ctx context.Context) (activeSupervisor, stand
 	}
 }
 
+type linkQualification struct {
```

Comment:
I'd suggest to move all of the link qualification code into a new file (`linkqual.go`) within the same package -- otherwise this `gnoi.go` file is going to get somewhat unmanageable.